src/server/routes/editor.js

```diff
@@ -357,170 +354,180 @@ router.get('/:id/revisions', async (req, res, next) => {
 	const {AchievementUnlock} = orm;
-	return new AchievementUnlock({
+
+	// Get the achievement unlock which is currently in this "rank"/slot
+	const unlockToUnrank = await new AchievementUnlock({
 		editorId,
 		profileRank: rank
 	})
-		.fetch({require: false})
-		.then((unlock) => {
-			if (unlock !== null) {
-				unlock.set('profileRank', null)
-					.save();
-			}
-		})
-		.then(() => {
-			let updatePromise;
-			if (bodyRank === '') {
-				updatePromise = new Promise(resolve => resolve(false));
-			}
-			else {
-				updatePromise = new AchievementUnlock({
-					achievementId: parseInt(bodyRank, 10),
-					editorId: parseInt(editorId, 10)
-				})
-					.fetch({require: true})
-					.then((unlock) => unlock.set('profileRank', rank).save());
-			}
-			return updatePromise;
-		});
+		.fetch({require: false});
+
+
+	// If there's a match, then unset the rank on the unlock, and save
+	if (unlockToUnrank !== null) {
+		await unlockToUnrank.set('profileRank', null).save();
+	}
+
+	if (bodyRank === '') {
+		return false;
+	}
+
+	// Then fetch the achievement to be placed in the specified rank
+	const unlockToRank = await new AchievementUnlock({
+		achievementId: parseInt(bodyRank, 10),
+		editorId
+	})
+		.fetch({require: true});
+
+	// TODO: this can throw, so missing error handling (but so was old code)
+
+	// And set the rank on the achievement and save it
+	return unlockToRank.set('profileRank', rank).save();
 }
 
 
 router.post('/:id/achievements/', auth.isAuthenticated, (req, res) => {
 	const {orm} = req.app.locals;
 	const {Editor} = orm;
 	const userId = parseInt(req.params.id, 10);
-	const editorPromise = new Editor({id: userId})
-		.fetch({
-			require: true,
-			withRelated: ['type', 'gender', 'area']
-		})
-		.then((editordata) => {
-			let editorJSON;
-
-			if (!req.user || userId !== req.user.id) {
-				editorJSON = new Promise((resolve, reject) => reject(new Error('Not authenticated')));
-			}
-			else {
-				editorJSON = new Promise(resolve => resolve(editordata.toJSON()));
-			}
-			return editorJSON;
-		});
 
-	const rankOnePromise = rankUpdate(orm, req.params.id, req.body.rank1, 1);
-	const rankTwoPromise = rankUpdate(orm, req.params.id, req.body.rank2, 2);
-	const rankThreePromise = rankUpdate(orm, req.params.id, req.body.rank3, 3);
-
-
-	const rankPromise =
-		editorPromise.then(() =>
-			Promise.all([
-				rankOnePromise,
-				rankTwoPromise,
-				rankThreePromise
-			]))
-			.then((rankJSON) => {
-				res.redirect(`/editor/${req.params.id}`);
-				return rankJSON;
-			});
-	handler.sendPromiseResult(res, rankPromise);
+	async function runAsync() {
+		if (!isCurrentUser(userId, req.user)) {
+			throw new Error('Not authenticated');
+		}
+
+		// Presumably, this is testing that the provided editor ID is valid
+		// prior to attempting to update the ranks. Possibly not needed since
+		// the user must exist if authenticated?
+		await new Editor({id: userId})
+			.fetch({require: true});
```

Comment:
I agree, I can't see any way for the user not to exist, and the `auth.isAuthenticated` middleware ensure that.
I think it can be removed.

Reply:
Removed in b2ec79a